src/main.rs

```diff
@@ -43,312 +44,301 @@ mod rpc;
-            }
-
-            x if x == MenuEntries::AddCursorAbove as u32 => {
-                self.s().edit_view.add_cursor_above(&self.win);
-            }
-            x if x == MenuEntries::AddCursorBelow as u32 => {
-                self.s().edit_view.add_cursor_below(&self.win);
-            }
-            x if x == MenuEntries::SingleSelection as u32 => {
-                self.s().edit_view.single_selection(&self.win);
-            }
-            x if x == MenuEntries::SelectAll as u32 => {
-                self.s().edit_view.select_all(&self.win);
-            }
-            _ => println!("unexpected id {}", id),
-        }
+impl App {
+    fn get_core(&self) -> std::sync::MutexGuard<'_, rpc::Core, > {
+        self.core.lock().unwrap()
     }
 
-    fn char(&self, ch: u32, mods: u32) {
-        self.s().edit_view.char(ch, mods, &self.win);
+    fn get_state(&self) -> std::sync::MutexGuard<'_, AppState, > {
+        self.state.lock().unwrap()
     }
 
-    fn keydown(&self, vk_code: i32, mods: u32) -> bool {
-        self.s().edit_view.keydown(vk_code, mods, &self.win)
-    }
-
-    fn mouse(&self, event: &MouseEvent) {
-        let (x_px, y_px) = self.win.handle.borrow().pixels_to_px_xy(event.x, event.y);
-        self.s().edit_view.mouse(x_px, y_px, event.mods, event.which, event.ty, &self.win);
-    }
-
-    fn mouse_wheel(&self, delta: i32, mods: u32) {
-        self.s().edit_view.mouse_wheel(delta, mods, &self.win)
-    }
-
-    fn destroy(&self) {
-        win_main::request_quit();
-    }
-
-    fn as_any(&self) -> &Any { self }
-}
-
-impl MainWinHandler {
-    /// Convenience function for accessing state.
-    fn s(&self) -> RefMut<MainWinState> {
-        self.win.state.borrow_mut()
+    fn get_handle(&self) -> std::sync::MutexGuard<'_, IdleHandle, > {
+        self.handle.lock().unwrap()
     }
 }
 
-impl MainWin {
+impl App {
     fn req_new_view(&self, filename: Option<&str>) {
         let mut params = json!({});
         if let Some(filename) = filename {
             params["file_path"] = json!(filename);
         }
-        let handle = self.handle.borrow().get_idle_handle().unwrap();
-        self.core.borrow_mut().send_request("new_view", &params,
+        let edit_view = 0;
+        let core = Arc::downgrade(&self.core);
+        let state = self.state.clone();
+        let handle = self.handle.clone();
+        self.core.lock().unwrap().send_request("new_view", &params,
             move |value| {
                 let value = value.clone();
-                handle.add_idle(move |a| {
-                    let handler = a.downcast_ref::<MainWinHandler>().unwrap();
-                    let edit_view = &mut handler.win.state.borrow_mut().edit_view;
-                    edit_view.set_view_id(value.as_str().unwrap());
-                });
+                let mut state = state.lock().unwrap();
+                let handle = handle.lock().unwrap();
+                state.focused = value.to_string();
```

Comment:
You don't want `to_string()` here, that'll format it as JSON. I think `.as_str().unwrap().to_string()`.